src/software/ai/navigator/navigator.cpp

```diff
@@ -2,303 +2,285 @@
+    for (const auto &intent : move_intents)
+    {
+        // start with non-MoveIntent robots and then add avoid areas
+        auto obstacles            = friendly_non_move_intent_robot_obstacles;
+        auto avoid_area_obstacles = getObstaclesFromAvoidAreas(intent.getAreasToAvoid());
+        obstacles.insert(obstacles.end(), avoid_area_obstacles.begin(),
+                         avoid_area_obstacles.end());
+
+        auto robot = world.friendlyTeam().getRobotById(intent.getRobotId());
+
+        if (robot)
         {
-            current_destination = path_points[1];
-            auto move           = std::make_unique<MovePrimitive>(
-                move_intent.getRobotId(), current_destination,
-                move_intent.getFinalAngle(),
-                move_intent.getFinalSpeed() *
-                    getCloseToEnemyObstacleFactor(path_points[1]),
-                move_intent.getDribblerEnable(), move_intent.getMoveType(),
-                move_intent.getAutoKickType());
-            current_primitive = std::move(move);
+            Point start = robot->position();
+            Point end   = intent.getDestination();
+
+            path_objectives.insert(PathObjective(start, end, robot->velocity().len(),
+                                                 obstacles, intent.getRobotId()));
         }
         else
         {
-            current_destination      = path_points[1];
-            double segment_final_vel = getCloseToEnemyObstacleFactor(
-                path_points[1] * calculateTransitionSpeedBetweenSegments(
-                                     path_points[0], path_points[1], path_points[2],
-                                     ROBOT_MAX_SPEED_METERS_PER_SECOND *
-                                         Util::DynamicParameters->getNavigatorConfig()
-                                             ->TransitionSpeedFactor()
-                                             ->value()));
-
-            auto move = std::make_unique<MovePrimitive>(
-                move_intent.getRobotId(), current_destination,
-                move_intent.getFinalAngle(), segment_final_vel,
-                move_intent.getDribblerEnable(), move_intent.getMoveType(),
-                move_intent.getAutoKickType());
-            current_primitive = std::move(move);
+            std::stringstream ss;
+            ss << "Failed to find robot associated with robot id = "
+               << intent.getRobotId();
+            LOG(WARNING) << ss.str();
         }
     }
-    else
-    {
-        LOG(WARNING) << "Path planner could not find a path";
-        auto stop = std::make_unique<StopPrimitive>(move_intent.getRobotId(), false);
-        current_primitive = std::move(stop);
-    }
+    return path_objectives;
 }
 
-std::optional<Obstacle> Navigator::obstacleFromAvoidArea(AvoidArea avoid_area)
+void Navigator::addMoveIntentsToAssignedPrimitives(
+    std::vector<std::unique_ptr<Primitive>> &assigned_primitives)
```

Comment:
Two questions here:
~1) Why are we passing in an empty vector by reference? Why not return it?~
~2) I'm confused about the name, I read this function as "~

Edit: Ah, I see what's going on here. We should really try to avoid in-place mutation like this, it makes the data flow harder to follow. I know it's a bit more work at the caller site, but could we change this to `std::vector<std::unique_ptr<Primitive>> Navigator::getPrimitivesForMoveIntents(std::vector<MoveIntent> move_intents)` (and propagate the `move_intents` down through functions, rather then grabbing them as class members?). 

The general concern I'm trying to address here is that there is a fair amount of indirection in this class (which is of course linked to our use of the `Visitor` pattern), and I really want to make it easy to trace the flow of control/data.

My personal rule for this sort of stuff is that if there is an obvious principle set of data a function is dealing with, it should be passed in through args. "Tertiary" stuff that you maybe wouldn't immediately think of the function needing can be accessed as class members. Obviously this is pretty vague/handwavy, but would be interested to get your thoughts!

Reply:
After some offline discussion, we plan to move a lot of functions that don't rely on Navigator member variables into a separate file

Reply:
I'm going to quickly chime in here and way that if this function gets changed to return values / do less mutating, can we change the name as well? I know what it does because I have lots of context about the Navigator, but think I wouldn't quite understand what this function did I if was reading this without context. Not going to suggest anything until changes are made, just putting it out there.

Reply:
Yup I've changed the names to be consistent with returning things